File: 2023/11/solution.js
const fs = require("fs");
const path = require("path");

const input = fs
  .readFileSync(path.join(__dirname, "input.txt"), "utf-8")
  .split("\n")
  .map((a) => a.split(""));

const emptyRows = [];
const emptyCols = [];
const galaxies = [];

for (let c = 0; c < input[0].length; c++) {
  let galaxyFound = false;
  for (let r = 0; r < input.length; r++) {
    if (input[r][c] === "#") {
      galaxies.push([r, c]);
      galaxyFound = true;
    }
  }
  if (!galaxyFound) emptyCols.push(c);
}

for (let r = 0; r < input.length; r++) {
  if (!input[r].includes("#")) emptyRows.push(r);
}

let part1 = 0;
let part2 = 0;

for (let i = 0; i < galaxies.length; i++) {
  for (let j = i + 1; j < galaxies.length; j++) {
    let [r1, c1] = galaxies[i];
    let [r2, c2] = galaxies[j];
    [r1, r2] = [Math.min(r1, r2), Math.max(r1, r2)];
    [c1, c2] = [Math.min(c1, c2), Math.max(c1, c2)];
    const expandedRows = emptyRows.filter((r) => r > r1 && r < r2).length;
    const expandedCols = emptyCols.filter((c) => c > c1 && c < c2).length;
    part1 += r2 - r1 + c2 - c1 + expandedRows + expandedCols;
    part2 +=
      r2 -
      r1 +
      c2 -
      c1 +
      (1000000 - 1) * expandedRows +
      (1000000 - 1) * expandedCols;
  }
}

console.log(part1);
console.log(part2);

File: 2023/11/solution.go
package main

import (
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"strings"

	"github.com/weitecklee/adventofcode/utils"
)

func main() {
	_, filename, _, _ := runtime.Caller(0)
	dirname := filepath.Dir(filename)
	inputFilePath := filepath.Join(dirname, "input.txt")
	data, err := os.ReadFile(inputFilePath)
	if err != nil {
		panic(err)
	}
	puzzleInput := strings.Split(string(data), "\n")
	fmt.Println(solve(puzzleInput, 2))
	fmt.Println(solve(puzzleInput, 1000000))
}

func solve(puzzleInput []string, expansionFactor int) int {
	var galaxies [][2]int
	var rowsToExpand, colsToExpand []int
	var expandRow, expandCol bool
	for r, row := range puzzleInput {
		expandRow = true
		for c, ch := range row {
			if ch == '#' {
				galaxies = append(galaxies, [2]int{r, c})
				expandRow = false
			}
		}
		if expandRow {
			rowsToExpand = append(rowsToExpand, r)
		}
	}
	for c := range puzzleInput[0] {
		expandCol = true
		for r := range puzzleInput {
			if puzzleInput[r][c] == '#' {
				expandCol = false
				break
			}
		}
		if expandCol {
			colsToExpand = append(colsToExpand, c)
		}
	}
	res := 0
	for i, galaxy1 := range galaxies {
		for _, galaxy2 := range galaxies[i+1:] {
			d := utils.AbsInt(galaxy1[0]-galaxy2[0]) + utils.AbsInt(galaxy1[1]-galaxy2[1])
			for _, r := range rowsToExpand {
				if (r > galaxy1[0] && r < galaxy2[0]) || (r < galaxy1[0] && r > galaxy2[0]) {
					d += expansionFactor - 1
				}
			}
			for _, c := range colsToExpand {
				if (c > galaxy1[1] && c < galaxy2[1]) || (c < galaxy1[1] && c > galaxy2[1]) {
					d += expansionFactor - 1
				}
			}
			res += d
		}
	}
	return res
}
